llvm/include/llvm/BitstreamCAS/BitstreamCASWriter.h

```diff
@@ -0,0 +1,685 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This header defines the BitstreamCASWriter class.  This class can be used to
+// write an arbitrary bitstream, regardless of its contents.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_BITSTREAM_BITSTREAMWRITER_H
+#define LLVM_BITSTREAM_BITSTREAMWRITER_H
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Bitstream/BitCodes.h"
+#include "llvm/BitstreamCAS/BitstreamCASNode.h"
+#include "llvm/CAS/CASReference.h"
+#include "llvm/CAS/ObjectStore.h"
+#include "llvm/Support/Endian.h"
+#include "llvm/Support/MathExtras.h"
+#include "llvm/Support/raw_ostream.h"
+#include <algorithm>
+#include <memory>
+#include <optional>
+#include <vector>
+
+namespace llvm {
+
+class BitstreamCASWriter {
+  /// Out - The buffer that keeps unflushed bytes.
+  SmallVectorImpl<char> &Out;
+
+  /// CAS - The object store that the bitstream is written to.
+  std::shared_ptr<cas::ObjectStore> &CAS;
+
+  /// FS - The file stream that Out flushes to. If FS is nullptr, it does not
+  /// support read or seek, Out cannot be flushed until all data are written.
+  std::shared_ptr<raw_fd_stream> &FS;
+
+  /// CurBit - Always between 0 and 31 inclusive, specifies the next bit to use.
+  unsigned CurBit;
+
+  /// CurValue - The current value. Only bits < CurBit are valid.
+  uint32_t CurValue;
+
+  /// CurCodeSize - This is the declared size of code values used for the
+  /// current block, in bits.
+  unsigned CurCodeSize;
+
+  /// BlockInfoCurBID - When emitting a BLOCKINFO_BLOCK, this is the currently
+  /// selected BLOCK ID.
+  unsigned BlockInfoCurBID;
+
+  /// CurAbbrevs - Abbrevs installed at in this block.
+  std::vector<std::shared_ptr<BitCodeAbbrev>> CurAbbrevs;
+
+  SmallVectorImpl<char> &Header;
+
+  std::stack<std::shared_ptr<cas::BitstreamCASNode>> NodeStack;
+
+  std::vector<cas::ObjectRef> TopLevelNodeRefs; // Think if we can remove this.
+
+  struct Block {
+    unsigned PrevCodeSize;
+    size_t StartSizeWord;
+    std::vector<std::shared_ptr<BitCodeAbbrev>> PrevAbbrevs;
+    Block(unsigned PCS, size_t SSW) : PrevCodeSize(PCS), StartSizeWord(SSW) {}
+  };
+
+  /// BlockScope - This tracks the current blocks that we have entered.
+  std::vector<Block> BlockScope;
+
+  /// BlockInfo - This contains information emitted to BLOCKINFO_BLOCK blocks.
+  /// These describe abbreviations that all blocks of the specified ID inherit.
+  struct BlockInfo {
+    unsigned BlockID;
+    std::vector<std::shared_ptr<BitCodeAbbrev>> Abbrevs;
+  };
+  std::vector<BlockInfo> BlockInfoRecords;
+
+  void WriteWord(unsigned Value) {
+    Value =
+        support::endian::byte_swap<uint32_t, llvm::endianness::little>(Value);
+    if (NodeStack.empty()){
+      Header.append(reinterpret_cast<const char *>(&Value),
```

Comment:
I don't understand what is `Header` and why WriteWord need to write to two locations when NodeStack is empty.

Reply:
Yeah, Initially, I added it to support the header emitted by ASTWriter before entering the block. Now I have updated the logic to write this to the node's data.